File: diagonal_traversal.cpp
vector<int> diagonal(Node *root)
{
    vector<int> ans;
    if(root == nullptr) return ans;
    queue<Node*> q;
    q.push(root);
    while(!q.empty()){
        Node* node = q.front();
        q.pop();
        while(node){
            if(node->left){
                q.push(node->left);
            }
            ans.push_back(node->data);
            node = node->right;
        }
    }
    return ans;
}

File: diagonal_traversal.java
class Tree
{
    public ArrayList<Integer> diagonal(Node root)
    {
        ArrayList<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        Queue<Node> q = new ArrayDeque<>();
        q.add(root);
        while(!q.isEmpty()){
            Node node = q.remove();
            while(node != null){
                ans.add(node.data);
                if(node.left != null) q.add(node.left);
                node = node.right;
            }
        }
        return ans;
    }
}
